### PR TITLE
NIFI-16159 Add support for FileResourceService in PutFile

### PR DESCRIPTION
**Pull request by shubhluck**
# Summary

[NIFI-16159](https://issues.apache.org/jira/browse/NIFI-16159)

Add support for `FileResourceService` in `PutFile`, matching the `Resource Transfer Source` / `File Resource Service` pattern already available in `PutS3Object` (NIFI-12642), `PutGCSObject` (NIFI-12643), and `PutHDFS` (NIFI-12801).

`PutFile` currently writes only the incoming FlowFile's content to the destination directory. Users who want to stage a file already on the local filesystem (or reachable via a custom `FileResourceService` implementation) have to first pull it into a FlowFile with `FetchFile`/`GetFile`, which is wasteful for large local files.

**Changes**

- Add `Resource Transfer Source` and `File Resource Service` property descriptors to `PutFile`, reusing `ResourceTransferProperties` from `nifi-resource-transfer`.
- When `Resource Transfer Source = File Resource Service`, obtain a `FileResource` via `ResourceTransferUtils.getFileResource(...)` and stream its `InputStream` to the `.<filename>` temp file with `StandardCopyOption.REPLACE_EXISTING` (preserving the overwrite semantics of the original `session.exportTo(...)` call), then proceed with the existing rename / permissions / owner / group / lastModified logic.
- When `Resource Transfer Source = FlowFile Content` (default) the behavior is byte-identical to today.
- Add `nifi-resource-transfer` and `nifi-file-resource-service-api` (compile) and `nifi-file-resource-service` (test) to `nifi-standard-processors/pom.xml`.
- Add `TestPutFile#testPutFileFromLocalFile` covering the `FileResourceService` path using `StandardFileResourceService`.

**Backward compatibility**

`Resource Transfer Source` defaults to `FlowFile Content`, so existing flows are unaffected. `File Resource Service` only appears in the UI when the source is set to `File Resource Service` (declared via `dependsOn`).

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`
- [x] Pull request contains [commits signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) with a registered key indicating `Verified` status

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `./mvnw clean install -P contrib-check`
  - [ ] JDK 21
  - [ ] JDK 25

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by shubhluck** [comment]
The `TestSocketLoadBalancedFlowFileQueue.testGetQueueSnapshotIncludesRebalancingPartition` failure on the JDK 21 EN leg is unrelated to this change (which only touches nifi-standard-processors/PutFile). The other three matrix legs passed. Looks like a flaky test.

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutFile.java [hunk 1]
```diff
@@ -304,7 +315,16 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
                 }
             }
 
-            session.exportTo(flowFile, dotCopyFile, false);
+            final ResourceTransferSource resourceTransferSource = ResourceTransferSource.valueOf(
+                    context.getProperty(RESOURCE_TRANSFER_SOURCE).getValue());
+            final Optional<FileResource> fileResource = getFileResource(resourceTransferSource, context, flowFile.getAttributes());
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutFile.java:319]: This should use `asAllowableValue()` instead of `ResourceTransferSource.valueOf()`

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml [hunk 2]
```diff
@@ -38,6 +38,21 @@
             <artifactId>nifi-standard-record-utils</artifactId>
             <version>2.11.0-SNAPSHOT</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-resource-transfer</artifactId>
+            <version>2.11.0-SNAPSHOT</version>
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: The versions should be updated to `2.12.0-SNAPSHOT`.

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFile.java [hunk 3]
```diff
@@ -244,6 +249,42 @@ public void testReplaceAndMaxFileLimitReach() throws IOException {
...
+        final FileResourceService service = new StandardFileResourceService();
+        runner.addControllerService(serviceId, service);
+        runner.setProperty(service, StandardFileResourceService.FILE_PATH, String.format("${%s}", attributeName));
+        runner.enableControllerService(service);
+
+        runner.setProperty(ResourceTransferProperties.RESOURCE_TRANSFER_SOURCE, ResourceTransferSource.FILE_RESOURCE_SERVICE.getValue());
+        runner.setProperty(ResourceTransferProperties.FILE_RESOURCE_SERVICE, serviceId);
+
+        final byte[] fileData = "0123456789".getBytes(StandardCharsets.UTF_8);
+        final Path tempFilePath = Files.createTempFile("PutFile_testPutFileFromLocalFile_", "");
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFile.java]: This approach can leave files in the temporary directory. The JUnit TempDir annotation should be used to wire in a directory to the test method instead.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for proposing this addition @shubhluck. Although using the File Resource Service for local transfer is a bit of an edge case, it is minimal enough that it looks straightforward. I noted a few minor recommendations, then this should be ready to go.

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutFile.java (hunk 1 above)
Comment (shubhluck) [reply nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutFile.java:319]: Fixed as suggested switched to asAllowableValue

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml (hunk 2 above)
Comment (shubhluck) [reply nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: Bumped to 2.12.0-SNAPSHOT.

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFile.java (hunk 3 above)
Comment (shubhluck) [reply nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFile.java]: Switched to a @TempDir Path sourceDir method parameter and dropped the try/finally cleanup

**Review by exceptionfactory — APPROVED** [review]
Thanks @shubhluck, the latest version looks good! +1 merging
